### ISSUE TITLE
ISSUE SCRAPING STOPS: AttributeError: 'NoneType' object has no attribute 'replace' 

### ISSUE THREAD

**Issue by Allacius** [twintproject/twint#293]
### Initial Check
> If the issue is a request please specify that it is a request in the title (Example: [REQUEST] more features). If this is a question regarding 'twint' please specify that it's a question in the title (Example: [QUESTION] What is x?). Please **only** submit issues related to 'twint'. Thanks.

>Make sure you've checked the following:

- [yep] Python version is 3.6;
- [?] Updated Twint with `pip3 install --upgrade -e git+https://github.com/twintproject/twint.git@origin/master#egg=twint`; --> Note to that: when I try this command it returns: "Did not find branch or tag 'origin/master', assuming revision or ref."
- [desperately] I have searched the issues and there are no duplicates of this issue/question/request.

### Command Ran
>python3 /Applications/twint/Twint.py -s “#westminsterbridge” --since 2017-03-22 --until 2017-03-23 -o westminsterbridge.csv --csv

### Description of Issue
The scraping stops with:

Traceback (most recent call last):
  File "/Users/AAA/src/twint/Twint.py", line 260, in <module>
    main()
  File "/Users/AAA/src/twint/Twint.py", line 252, in main
    twint.run.Search(c)
  File "/Users/AAA/src/twint/twint/run.py", line 215, in Search
    run(config)
  File "/Users/AAA/src/twint/twint/run.py", line 167, in run
    get_event_loop().run_until_complete(Twint(config).main())
  File "/Library/Frameworks/Python.framework/Versions/3.6/lib/python3.6/asyncio/base_events.py", line 468, in run_until_complete
    return future.result()
  File "/Users/AAA/src/twint/twint/run.py", line 135, in main
    await self.tweets()
  File "/Users/AAA/src/twint/twint/run.py", line 117, in tweets
    await output.Tweets(tweet, "", self.config, self.conn)
  File "/Users/AAA/src/twint/twint/output.py", line 116, in Tweets
    await checkData(tweets, location, config, conn)
  File "/Users/AAA/src/twint/twint/output.py", line 90, in checkData
    if datecheck(tweet.datestamp, config):
  File "/Users/AAA/src/twint/twint/output.py", line 25, in datecheck
    d = int(datestamp.replace("-", ""))
AttributeError: 'NoneType' object has no attribute 'replace' 

Any hint? I've updated up to this version after I've got stuck with a loop-error with any kind of search: Twint got stuck every time in a loop of "n" tweet that it kept repeting. Now it seems to go as usual but all of a sudden stops and display that error message. Any Hint? I reeeeeeallly appreciate your passion and effort! Thanks! 

### Environment Details
>Using Windows, Linux? What OS version? Running this in Anaconda? Jupyter Notebook? Terminal?
Using Mac OS 10.13.6

**Comment by Allacius** [comment]
I can add that until I scrape a single user or perform a "-s" search *without* any date specification it works really well, without any error.

**Comment by pielco11** [comment]
I got the same error and Twint stops at a specific tweet. Searching with Twitter Search I found that after that tweet there is a message that says "@.... account is temporary suspended..." and then other tweets are visualized, so I think that a tweet exists at the point, but no infos are provided and that's why Twint stops.

Plus, resuming the scraping from the first (after "none"-tweet) available tweet which id is `844606628511059969`, everything keeps going as usual. The query to achieve this is `python3 Twint.py -s “#westminsterbridge” --since 2017-03-22 --until 2017-03-23 -o westminsterbridge.csv --csv --resume 844606628511059969`

In conclusion, until we do not push a patch to handle "none-tweets" I suggest you to use Twitter Search to verify that the error is caused by a none-tweet and use the next available tweet-id to resume the scraping session

**Comment by Allacius** [comment]
Hello, thank you for your suggestion! 
If I've understood correctly: you perform the exact same query in twitter search, then you scroll down untill you have found the last tweet retrieved by twint, and then you have the ID tweet to start with. 
Is it correct?  
I will try it now

**Comment by Allacius** [comment]
OK, so the scrolling scrollin will be a mitigation (untill your goodwill won't save my spacebar, at least). 
Instead, I wasn't able to use you workaround: I have this sequence of ID tweets: 
844598713003757569
844598710986326022
and then twint stops. 
if I perform the same search with the suggested --resume 844598710986326023 (last tweet id +1) Twint give back the same error.. :(

**Comment by pielco11** [comment]
My bad, I was wrong... since we are going in the past from the present, we have to decrease the id value:
```python
import twint

c = twint.Config()

c.Search = "#westminsterbridge"
c.Store_csv = True
c.Store_object = True
c.Output = "westminsterbridge"
c.Since = "2017-03-22"
c.Until = "2017-03-23"
i = 0

while True:
  try:
    twint.run.Search(c)
  except AttributeError:
    i += 100000
    tweets = twint.output.tweets_object
    c.Resume = tweets[len(tweets) - 1].id - i
    print("broken id :: " + str(c.Resume))
```

I setted `i += 100000` because there are a lot of tweets and a granular search could require too much time

**Comment by Allacius** [comment]
> My bad, I was wrong... since we are going in the past from the present, we have to decrease the id value:
> 
> ```python
> import twint
> 
> c = twint.Config()
> 
> c.Search = "#westminsterbridge"
> c.Store_csv = True
> c.Store_object = True
> c.Output = "westminsterbridge"
> c.Since = "2017-03-22"
> c.Until = "2017-03-23"
> i = 0
> 
> while True:
>   try:
>     twint.run.Search(c)
>   except AttributeError:
>     i += 100000
>     tweets = twint.output.tweets_object
>     c.Resume = tweets[len(tweets) - 1].id - i
>     print("broken id :: " + str(c.Resume))
> ```
> 
> I setted `i += 100000` because there are a lot of tweets and a granular search could require too much time
> 
> **UPDATE**:
> with `i += 10000000000` in a couple of tries Twint restarts correctly

This afternoon I've tryed the first option, but I had the impression that Twint was stuck in a sort of never-ending list of "broken id ::" that went over for more than 3 hrs.. I was going to try that again, leaving twint scraping all nitght, but now I will give a try with your last update! 
I'll let you know, again thanks for the precious hard work!

**Comment by pielco11** [comment]
I'm going to put down my karma points for that embarrassing error

**Comment by Allacius** [comment]
Oh, gosh. So I've let this run all night: 
import twint

c = twint.Config()

c.Search = "#westminsterbridge"
c.Store_csv = True
c.Store_object = True
c.Output = "westminsterbridge"
c.Since = "2017-03-22"
c.Until = "2017-03-23"
i = 0

while True:
  try:
    twint.run.Search(c)
  except AttributeError:
    i += 10000000000
    tweets = twint.output.tweets_object
    c.Resume = tweets[len(tweets) - 1].id - i
    print("broken id :: " + str(c.Resume))

But then I woke up with this: 

Traceback (most recent call last):
  File "/Users/AAA/src/twint/twint/get.py", line 105, in Response
    async with session.get(url, ssl=False, params=params) as response:
  File "/Library/Frameworks/Python.framework/Versions/3.6/lib/python3.6/site-packages/aiohttp/client.py", line 843, in __aenter__
    self._resp = await self._coro
  File "/Library/Frameworks/Python.framework/Versions/3.6/lib/python3.6/site-packages/aiohttp/client.py", line 366, in _request
    timeout=timeout
  File "/Library/Frameworks/Python.framework/Versions/3.6/lib/python3.6/site-packages/aiohttp/connector.py", line 445, in connect
    proto = await self._create_connection(req, traces, timeout)
  File "/Library/Frameworks/Python.framework/Versions/3.6/lib/python3.6/site-packages/aiohttp/connector.py", line 757, in _create_connection
    req, traces, timeout)
  File "/Library/Frameworks/Python.framework/Versions/3.6/lib/python3.6/site-packages/aiohttp/connector.py", line 862, in _create_direct_connection
    req=req, client_error=client_error)
  File "/Library/Frameworks/Python.framework/Versions/3.6/lib/python3.6/site-packages/aiohttp/connector.py", line 822, in _wrap_create_connection
    return await self._loop.create_connection(*args, **kwargs)
  File "/Library/Frameworks/Python.framework/Versions/3.6/lib/python3.6/asyncio/base_events.py", line 765, in create_connection
    yield from self.sock_connect(sock, address)
  File "/Library/Frameworks/Python.framework/Versions/3.6/lib/python3.6/asyncio/selector_events.py", line 450, in sock_connect
    return (yield from fut)
concurrent.futures._base.CancelledError

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/Users/AAA/src/twint/TESTSCRAPER.py", line 15, in <module>
    twint.run.Search(c)
  File "/Users/AAA/src/twint/twint/run.py", line 215, in Search
    run(config)
  File "/Users/AAA/src/twint/twint/run.py", line 167, in run
    get_event_loop().run_until_complete(Twint(config).main())
  File "/Library/Frameworks/Python.framework/Versions/3.6/lib/python3.6/asyncio/base_events.py", line 468, in run_until_complete
    return future.result()
  File "/Users/AAA/src/twint/twint/run.py", line 135, in main
    await self.tweets()
  File "/Users/AAA/src/twint/twint/run.py", line 111, in tweets
    await self.Feed()
  File "/Users/AAA/src/twint/twint/run.py", line 41, in Feed
    response = await get.RequestUrl(self.config, self.init, headers=[("User-Agent", self.user_agent)])
  File "/Users/AAA/src/twint/twint/get.py", line 59, in RequestUrl
    response = await Request(_url, params=params, connector=_connector, headers=headers)
  File "/Users/AAA/src/twint/twint/get.py", line 100, in Request
    return await Response(session, url, params)
  File "/Users/AAA/src/twint/twint/get.py", line 106, in Response
    return await response.text()
  File "/Library/Frameworks/Python.framework/Versions/3.6/lib/python3.6/site-packages/async_timeout/__init__.py", line 45, in __exit__
    self._do_exit(exc_type)
  File "/Library/Frameworks/Python.framework/Versions/3.6/lib/python3.6/site-packages/async_timeout/__init__.py", line 92, in _do_exit
    raise asyncio.TimeoutError
concurrent.futures._base.TimeoutError

Any clue? It seems to me a different kind of problem.. but here stops my ability to understand what..

**Comment by pielco11** [comment]
It could be that Twitter has blocked some of your requests (this issue is concordant with Twitter TOS, so it's not something related to not working code), may you confirm that you got more tweets than before the proposed solution? (this is what I got, but I want to be sure)

**Comment by Allacius** [comment]
> It could be that Twitter has blocked some of your requests (this issue is concordant with Twitter TOS, so it's not something related to not working code), may you confirm that you got more tweets than before the proposed solution? (this is what I got, but I want to be sure)

Yes, indeed! Moreover, with the ".resume" line I am able to continue the scraping.. I think that timeout is somehow connected to the speed of the available internet connection. 

Anyway, @pielco11 thank you so much! I was able to download a much wider dataset and that's just great! 

I have one more question, but I think that could fit in a new thread: prior of the twint update, I was able to query by indicating only LAT and LONG and scraping all the tweet happening in an area of interest, indipendently from their content. If I try it now, Twint ask at least for an -s or -u arguments. Will you guys think of introduce that feature again? Or does it depend on some changes from Twitter? 
Thanks again and kudos to you!

**Comment by pielco11** [comment]
That was an issue with args checking, I tested and now it should be fine

**Attention here**
Please note that now Twint is **entirely** a module, if you want to install your local clone repo you have to do `pip3 install git+file:///path/to/repo@master` (see [here for windows](https://blogs.msdn.microsoft.com/ie/2006/12/06/file-uris-in-windows/))

**Comment by flyfisher123** [comment]
Is the --resume - workaround still working for you? I'm trying to scape all tweets containing "gdp" and all of a sudden it restarts from the latest tweets even if I try to resume from a specific id. The same query was working fine a few hours ago.

I have the same issues when I try the script from above to automate resume after an error.

**Comment by pielco11** [comment]
Maybe `i` is too big and Twitter restarts... if there is an issue, it's present with every similar query and it should not be present only with specific keywords

**Comment by pielco11** [comment]
@flyfisher123 may you try using `c.Until` and `c.Since`? Something like that when Twint stops, you change those fields to restart the scraping and getting new tweets. Please note that there will be some duplicates that can be easily removed

**Comment by flyfisher123** [comment]
@pielco11 Thanks a lot for your help.
Here's what I did (I've chosen a different keyword with less tweets for now):
`python3 Twint.py -s "HICP" -o HICP.csv --csv -l en --count --stats`
that gets me the message that 6553 tweets were succesfully collected. The last tweet is [this tweet](https://twitter.com/locationcyprus/status/651790159491563520) from 2015-10-07 18:04:18 CET, so that's obviously not all there is.
I've been scraping tweets for similar keywords for the last days and when this happened before, I always resumed with a slightly lower id number. So I tried that with `--resume 651590159491563520` and it just starts again from today.
If I try `--until 2015-10-08` it only collects the one tweet mentioned above.
With `--until 2015-10-07` it doesn't find any tweets and with `--until 2015-10-06` it starts up again and finds 15875 more tweets in between 2015-10-02 and 2008-02-29 which should be all of them.
Missing in between are those two tweets: [one](https://twitter.com/FinanzLinksWELT/status/651331815005786112), [two](https://twitter.com/CYSTAT_EN/status/651323626772258816).

**Comment by pielco11** [comment]
What you are getting with this query seems to be correct and not issue with Twitn

### only one tweet
![immagine](https://user-images.githubusercontent.com/17363750/48915544-576fc480-ee7f-11e8-9dc1-21338f42bbda.png)

### nothing found
![immagine](https://user-images.githubusercontent.com/17363750/48915560-65bde080-ee7f-11e8-904d-7abf540d6d12.png)

### other tweets
![immagine](https://user-images.githubusercontent.com/17363750/48915580-753d2980-ee7f-11e8-8f7f-e63842cc97c1.png)

So I guess that with `--until` Twitter does not go back too much, like a couple of days. I would suggest you to:

- use `--resume` if you find a tweet that can not be scraped, this happened before;
- iterate using `--since` and `--until` if no tweets were found.

If you get errors that you would not expect, use Twitter Advanced Search and verify that what you are getting correspond what Twitter gives you. If you find tweets via search and you are not scraping them, feel free to notify me and we will find a solution!

**Comment by flyfisher123** [comment]
Thanks a lot. I'll let you know over the next days.

**Comment by flyfisher123** [comment]
@pielco11 Any idea why resume isn't working? It always restarts. I tried the example from above with westminsterbridge and as seen in the screenshot it just starts from the beginning and the tweed id is higher than the resume id:
<img width="1244" alt="bildschirmfoto 2018-11-24 um 15 50 58" src="https://user-images.githubusercontent.com/45214435/48969696-86fb0a00-f002-11e8-93ea-23dc1b816917.png">

**Comment by pielco11** [comment]
@flyfisher123 maybe Twitter changed something on its side, I tried with a couple of queries and it seems that `--resume` is broken

**Comment by Allacius** [comment]
Hello @pielco11, after few days, I'm having trouble again with a loop in tweet retrieved. I perform this query: 

import twint

c = twint.Config()

c.Search = "London attack OR #londonattack"
c.Store_csv = True
c.Store_object = True
c.Output = "AAA"
c.Since = "2017-06-03"
c.Until = "2017-06-04"

i = 0

while True:
  try:
    twint.run.Search(c)
  except AttributeError:
    i += 100000000000000000
    tweets = twint.output.tweets_object
    c.Resume = tweets[len(tweets) - 1].id - i
    print("broken id :: " + str(c.Resume))

But then I'm trapped in a loop of only few tweets ranging from 01:29:13 of june 4 until  1:59:59 of june 4. 
Any hint? I'm stuck without a clue.. I've tryed to reduce the number of i but it doesn't solved my problem.. 
Thanks!

**Comment by pielco11** [comment]
Hello @Allacius ,

it seems that `Resume` is broken now, it was working great days ago. I guess that Twitter changed something. Plus users are reporting issues also while scraping data, specifically Twitter is limiting the number of tweets that we can scrape.

Here is a workaround with a couple of downsides (have to wait and duplicated tweets)
```python
import time
import twint
 
c = twint.Config()
 
c.Search = "#metoo"
c.Store_csv = True
c.Store_object = True
c.Output = "output_directory"
 
_max_cycles = 100
_wait = 3*60 # 3 minutes
 
while _max_cycles:
  twint.run.Search(c)
  tweets = twint.output.tweets_object
  last_tweet = tweets[len(tweets) - 1]
  last_day = last_tweet.datestamp
  c.Until = last_day
  _max_cycles -= 1
  time.sleep(_wait)
```

**Comment by Allacius** [comment]
Who would win? Twitter or our Workaround? 
Thanks @pielco11,  I'll guess I'll wait then!

**Comment by pielco11** [comment]
@Allacius I edited your comment, the snippet should work

**Comment by pielco11** [comment]
Replace `twint.run.Search(c)` with this snippet (please be careful since this will catch **every** `AttributeError`)

```python
try:
  twint.run.Search(c)
except AttributeError:
  print("\n[!] found a removed tweet, probably\n")
```

**Comment by Allacius** [comment]
yep, it happened, it says: "[!] found a removed tweet, probably" and then stopped just at after few tweets retrieved.

**Comment by tdicksteinpomona** [comment]
Is c.Resume working again, or has anyone thought of a good way to resume right from a removed tweet? I'm trying to scrape a somewhat large (~100M tweets) dataset, and would rather not skip to the next day as soon as I encounter a broken/removed tweet

**Comment by pielco11** [comment]
@tdicksteinpomona you have just to handle the exception

**Comment by EonYang** [comment]
I temporarily solved this in my machine, by modifying this block in twint/output.py  line 23 :
```python
## what it was
def datecheck(datestamp, config):
    #logging.info("[<] " + str(datetime.now()) + ':: output+datecheck')
    if config.Since and config.Until:
        d = int(datestamp.replace("-", ""))
        s = int(config.Since.replace("-", ""))
        if d < s:
            return False
    return True
```

What I've added :
```python
if datestamp is None:
        return False
```

The new look:
```python
def datecheck(datestamp, config):
    #logging.info("[<] " + str(datetime.now()) + ':: output+datecheck')
    if datestamp is None:
        return False
    if config.Since and config.Until:
        d = int(datestamp.replace("-", ""))
        s = int(config.Since.replace("-", ""))
        if d < s:
            return False
    return True
```

**Comment by pielco11** [comment]
Hi @EonYang and thank you for the tip!

I guess that this is a good starting point but we can do something more to handle this kind of errors and help the user to understand what is going on

**Comment by matis28** [comment]
Hey!

For those who also are struggling with this error, but have implemented Twint as a package, an easy way out is to do the following in the class/file where twint.run.Search is executed:

`
import twint

def datecheck(datestamp, config):
    # logging.info("[<] " + str(datetime.now()) + ':: output+datecheck')
    if datestamp is None:
        return False
    if config.Since and config.Until:
        d = int(datestamp.replace("-", ""))
        s = int(config.Since.replace("-", ""))
        if d < s:
            return False
    return True


twint.output.datecheck = datecheck
`

Just wanted to share this as I just had the very same issue, maybe someone finds it useful.

Cheers!
